### PR TITLE
Get tuples working in LLVM

### PR DESCRIPTION
**Pull request by namannimmo10**

### CODE REVIEW

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -1731,6 +1733,9 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
                 llvm_type = list_type;
                 break;
             }
+            case (ASR::ttypeType::Tuple) : {
+                llvm_type = tuple_type;
```
Comment (namannimmo10) [inline src/libasr/codegen/asr_to_llvm.cpp:1737]: First, I am trying to make this simple test work: 
```python
def f():
    t: tuple[f64, f64]
    t = (43.5, 647.75)


f()
```

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 2]
```diff
@@ -1010,6 +1011,7 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
         complex_type_8 = llvm::StructType::create(context, els_8, "complex_8");
         complex_type_4_ptr = llvm::StructType::create(context, els_4_ptr, "complex_4_ptr");
         complex_type_8_ptr = llvm::StructType::create(context, els_8_ptr, "complex_8_ptr");
+        tuple_type = llvm::StructType::create(context, els_8, "tuple");
```
Comment (czgdp1807) [inline src/libasr/codegen/asr_to_llvm.cpp:1014]: `els_8` is a two element vector of pointer to double types. So that means all the tuples would look like `{double*, double*}`?

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 1 above)
Comment (czgdp1807) [reply src/libasr/codegen/asr_to_llvm.cpp:1737]: Another simple test,

```
t: tuple[f64, i32, str]
t = (43.5, 0, "a")
```
Making these two work simultaneously would be better.

**Comment by czgdp1807** [comment]
@namannimmo10 Are you still working on this?

**Comment by namannimmo10** [comment]
I think we're pretty close here. Does anyone know what's the issue:

```console
(lp) namannimmo@srf lpython % cat t.py
def f():
    t: tuple[i32, i32]
    t = (1, 2)



f()
(lp) namannimmo@srf lpython % lpython t.py
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/namannimmo/Applications/oss/lpython/src/bin/lpython.cpp", line 883
    err = compile_python_to_object_file(arg_file, tmp_o, runtime_library_dir, lpython_pass_manager, compiler_options, time_report);
  File "/Users/namannimmo/Applications/oss/lpython/src/bin/lpython.cpp", line 368
    res = fe.get_llvm3(*asr, pass_manager, diagnostics);
  File "/Users/namannimmo/Applications/oss/lpython/src/lpython/python_evaluator.cpp", line 57
    e->get_context(), al, lpm, compiler_options.platform,
  File "/Users/namannimmo/Applications/oss/lpython/src/libasr/codegen/asr_to_llvm.cpp", line 5256
    v.nested_func_types = pass_find_nested_vars(asr, context,
  File "../libasr/asr.h", line 3691
  File "../libasr/asr.h", line 3669
  File "../libasr/asr.h", line 3692
  File "../libasr/asr.h", line 3470
  File "/Users/namannimmo/Applications/oss/lpython/src/libasr/codegen/asr_to_llvm.cpp", line 1072
    visit_symbol(*item.second);
  File "../libasr/asr.h", line 3694
  File "../libasr/asr.h", line 3480
  File "/Users/namannimmo/Applications/oss/lpython/src/libasr/codegen/asr_to_llvm.cpp", line 2310
    visit_procedures(x);
  File "/Users/namannimmo/Applications/oss/lpython/src/libasr/codegen/asr_to_llvm.cpp", line 2384
    this->visit_stmt(*x.m_body[i]);
  File "../libasr/asr.h", line 3708
  File "../libasr/asr.h", line 3500
  File "/Users/namannimmo/Applications/oss/lpython/src/libasr/codegen/asr_to_llvm.cpp", line 2953
    ASR::ttype_t* value_type = ASRUtils::expr_type(x.m_value);
  File "../libasr/asr_utils.h", line 992
  File "../libasr/asr_utils.h", line 986
LCompilersException: Not implemented.
```

**Comment by czgdp1807** [comment]
> I think we're pretty close here

I am not sure about that. Can you add some good tests to make sure your changes/additions work as expected. Like the test should involve initialising tuples, reading from tuples and checking via assert that those values match. As of now I don't see any test in the diff? Am I missing something here?


> Does anyone know what's the issue:

If you can then add a print statement just before the exception to see which `exprType` is not handled in the switch case. You can also use a debugger to add break point at the line just before the exception to see which `exprType` is not handled in the swith case.

File: integration_tests/test_tuples.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+from ltypes import i32, f64
+
+def f():
+    t1: tuple[i32, f64, str, bool]
+    t1 = (1, 2.0, "3", True)
+    assert t1[0] == 1
+    assert t1[2] == "3"
+    assert t1[3]
```
Comment (czgdp1807) [inline integration_tests/test_tuples.py:8]:
```suggestion
    t1: tuple[i32, f64, str, bool]
    t2: tuple[i32, f64, str, bool]
    t1 = (1, 2.0, "3", True)
    t2 = (2, 3.0, "3", False)
    assert t2[0] - t1[0] == 1
    assert abs(t2[1] - t1[1] - 1) <= 1e-12 
    assert t1[2] == t2[2]
    assert t1[3] || t2[3]
```

File: integration_tests/test_tuples.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+    t2: tuple[i32, i32]
+    t2 = (1, 2)
+    a: i32
+    b: i32
+    (a, b) = t2
+    assert a == 1
+    assert b == 2
```
Comment (czgdp1807) [inline integration_tests/test_tuples.py:16]:
```suggestion
    t3: tuple[i32, i32]
    t4: tuple[f64, f64]
    t2 = (1, 2)
    t4 = (1.0, 2.0)
    a: i32
    b: i32
    af: f64
    bf: f64
    (a, b) = t3
    (af, bf) = t4
    assert af == a
    assert bf == b
```

**Comment by czgdp1807** [comment]
There's not much in my suggestions but I am trying to test a bunch of corner cases here. I am having a feeling (not sure as I haven't tested your code yet) that whenever two tuples with exactly the same type will be created, LLVM will create two structs of different types (like by using `.1`, `.2` suffices in the name even though the content of these types will be the same), which will make reading LLVM code a bit harder. We can easily solve this problem by caching the tuple types and re-using them when some other tuple of same type will be created.

**Comment by namannimmo10** [comment]
TODO:

- [ ] Fix the bug and make sure we're able to initialize tuples.
- [ ] Implement TupleItem visitor

**Comment by czgdp1807** [comment]
How to fix? - https://github.com/lcompilers/lpython/pull/821#issuecomment-1196857215

```diff
diff --git a/src/libasr/asr_utils.h b/src/libasr/asr_utils.h
index 84615e8af..80284e2b0 100644
--- a/src/libasr/asr_utils.h
+++ b/src/libasr/asr_utils.h
@@ -982,8 +982,13 @@ inline int extract_dimensions_from_ttype(ASR::ttype_t *x,
             m_dims = nullptr;
             break;
         }
+        case ASR::ttypeType::Tuple: {
+            n_dims = 0;
+            m_dims = nullptr;
+            break;
+        }
         default:
-            throw LCompilersException("Not implemented.");
+            throw LCompilersException("Not implemented " + ASRUtils::type_to_str_py
thon(x));
     }
     return n_dims;
 }
```

**Comment by czgdp1807** [comment]
@namannimmo10 Would it be okay for you if I take over the addition of tuples in LPython? Let me know.

**Comment by namannimmo10** [comment]
Go ahead!

**Comment by czgdp1807** [comment]
Closing as its getting done in https://github.com/lcompilers/lpython/pull/851
